File: src/routes/estates.js
const express = require('express');
const { v4: uuidv4 } = require('uuid');
const db = require('../config/database');
const { authenticate, authorize } = require('../middleware/auth');
const router = express.Router();

// Get all estates
router.get('/', authenticate, async (req, res) => {
  try {
    const result = await db.query(
      `SELECT e.*, c.name as company_name 
       FROM estates e 
       JOIN companies c ON e.company_id = c.id 
       WHERE e.company_id = $1 
       ORDER BY e.created_at DESC`,
      [req.user.company_id]
    );
    res.json({ estates: result.rows });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// Get single estate with phases
router.get('/:id', authenticate, async (req, res) => {
  try {
    const estateResult = await db.query(
      'SELECT * FROM estates WHERE id = $1 AND company_id = $2',
      [req.params.id, req.user.company_id]
    );

    if (estateResult.rows.length === 0) {
      return res.status(404).json({ error: 'Estate not found' });
    }

    const phasesResult = await db.query(
      'SELECT * FROM phases WHERE estate_id = $1 ORDER BY phase_number',
      [req.params.id]
    );

    res.json({ 
      estate: estateResult.rows[0],
      phases: phasesResult.rows 
    });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// Create estate
router.post('/', authenticate, authorize('super_admin', 'admin'), async (req, res) => {
  try {
    const { name, location, total_phases = 1 } = req.body;
    const id = uuidv4();

    const result = await db.query(
      `INSERT INTO estates (id, company_id, name, location, total_phases) 
       VALUES ($1, $2, $3, $4, $5) RETURNING *`,
      [id, req.user.company_id, name, location, total_phases]
    );

    // Auto-create phases
    for (let i = 1; i <= total_phases; i++) {
      await db.query(
        `INSERT INTO phases (id, estate_id, phase_number, phase_name) 
         VALUES ($1, $2, $3, $4)`,
        [uuidv4(), id, i, `Phase ${i}`]
      );
    }

    res.status(201).json({ estate: result.rows[0], message: 'Estate created with phases' });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// Update estate
router.put('/:id', authenticate, authorize('super_admin', 'admin'), async (req, res) => {
  try {
    const { name, location, status } = req.body;
    const result = await db.query(
      `UPDATE estates SET name = $1, location = $2, status = $3 
       WHERE id = $4 AND company_id = $5 RETURNING *`,
      [name, location, status, req.params.id, req.user.company_id]
    );

    if (result.rows.length === 0) {
      return res.status(404).json({ error: 'Estate not found' });
    }

    res.json({ estate: result.rows[0] });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// Delete estate
router.delete('/:id', authenticate, authorize('super_admin'), async (req, res) => {
  try {
    const result = await db.query(
      'DELETE FROM estates WHERE id = $1 AND company_id = $2 RETURNING *',
      [req.params.id, req.user.company_id]
    );

    if (result.rows.length === 0) {
      return res.status(404).json({ error: 'Estate not found' });
    }

    res.json({ message: 'Estate deleted successfully' });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

module.exports = router;

File: app/routers/estates.py
"""
==============================================================================
SMART BIN LTD - ESTATES ROUTER
==============================================================================
Replaces: src/routes/estates.js

ENDPOINTS:
  GET  /api/estates/          → List all estates for logged-in company
  GET  /api/estates/{id}      → Get single estate with its phases
  POST /api/estates/          → Create new estate (admin only)
  PUT  /api/estates/{id}      → Update estate (admin only)
  DELETE /api/estates/{id}    → Delete estate (super_admin only)

BUSINESS LOGIC:
- Every estate belongs to ONE company (multi-tenant isolation)
- Creating an estate auto-creates its phases (e.g., 4 phases = Phase 1-4)
- Only super_admin and admin can create/modify
- Clerks can only view
==============================================================================
"""

import uuid
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from app.database import get_db
from app.models import Estate, Phase
from app.schemas import EstateCreate, EstateUpdate, EstateOut, PhaseOut
from app.dependencies import get_current_user, require_role

router = APIRouter(prefix="/api/estates", tags=["estates"])


@router.get("/", response_model=list[EstateOut])
def list_estates(db: Session = Depends(get_db), user=Depends(get_current_user)):
    """
    List all estates for the logged-in user's company.
    Automatically filters by company_id — clerks can't see other companies' data.
    """
    return db.query(Estate).filter(
        Estate.company_id == user.company_id
    ).order_by(Estate.created_at.desc()).all()


@router.get("/{estate_id}", response_model=EstateOut)
def get_estate(estate_id: str, db: Session = Depends(get_db), user=Depends(get_current_user)):
    """
    Get a single estate with all its phases nested inside.
    Returns 404 if estate doesn't exist or belongs to another company.
    """
    estate = db.query(Estate).filter(
        Estate.id == estate_id,
        Estate.company_id == user.company_id
    ).first()
    
    if not estate:
        raise HTTPException(status_code=404, detail="Estate not found")
    
    # Load phases manually and attach to response
    phases = db.query(Phase).filter(Phase.estate_id == estate_id).order_by(Phase.phase_number).all()
    estate.phases = [PhaseOut.model_validate(p) for p in phases]
    return estate


@router.post("/", response_model=EstateOut)
def create_estate(req: EstateCreate, db: Session = Depends(get_db),
                  user=Depends(require_role("super_admin", "admin"))):
    """
    Create a new estate and auto-generate its phases.
    
    Example request:
    {
        "name": "GAKINDU COURT",
        "location": "Utawala, Nairobi",
        "total_phases": 4
    }
    
    This creates:
    - 1 estate row
    - 4 phase rows (Phase 1, Phase 2, Phase 3, Phase 4)
    """
    # Create the estate
    estate = Estate(
        id=str(uuid.uuid4()),
        company_id=user.company_id,
        name=req.name,
        location=req.location,
        total_phases=req.total_phases
    )
    db.add(estate)
    db.commit()
    db.refresh(estate)
    
    # Auto-create phases (1 to total_phases)
    for i in range(1, req.total_phases + 1):
        phase = Phase(
            id=str(uuid.uuid4()),
            estate_id=estate.id,
            phase_number=i,
            phase_name=f"Phase {i}"
        )
        db.add(phase)
    
    db.commit()
    db.refresh(estate)
    return estate


@router.put("/{estate_id}", response_model=EstateOut)
def update_estate(estate_id: str, req: EstateUpdate, db: Session = Depends(get_db),
                  user=Depends(require_role("super_admin", "admin"))):
    """
    Partial update — only send fields you want to change.
    """
    estate = db.query(Estate).filter(
        Estate.id == estate_id,
        Estate.company_id == user.company_id
    ).first()
    
    if not estate:
        raise HTTPException(status_code=404, detail="Estate not found")
    
    # Update only provided fields
    for field, value in req.model_dump(exclude_unset=True).items():
        setattr(estate, field, value)
    
    db.commit()
    db.refresh(estate)
    return estate


@router.delete("/{estate_id}")
def delete_estate(estate_id: str, db: Session = Depends(get_db),
                  user=Depends(require_role("super_admin"))):
    """
    Only super_admin can delete estates (dangerous operation).
    Cascade delete removes all phases, houses, residents, bills automatically.
    """
    estate = db.query(Estate).filter(
        Estate.id == estate_id,
        Estate.company_id == user.company_id
    ).first()
    
    if not estate:
        raise HTTPException(status_code=404, detail="Estate not found")
    
    db.delete(estate)
    db.commit()
    return {"message": "Estate deleted successfully"}
